### PR TITLE
Add `@param-out` support

### PR DESCRIPTION
**Pull request by staabm**
requires https://github.com/phpstan/phpdoc-parser/pull/150
refs https://github.com/phpstan/phpstan/issues/6426#issuecomment-1073791565

closes https://github.com/phpstan/phpstan/issues/7231
closes https://github.com/phpstan/phpstan/issues/6871
closes https://github.com/phpstan/phpstan/issues/6186
closes https://github.com/phpstan/phpstan/issues/4372

see https://psalm.dev/docs/annotating_code/supported_annotations/#param-out-psalm-param-out

just put togehter the basic building blocks and tests:
- [x] wire ResolvedPhpDocBlock
- [x] wire ResolvedDocNodeResolver
- [x] validate param-out phpdoc tags and tests
- [x] initial NodeScopeResolverTests
- [x] implement param-out type specification

I had the idea of adding param-out-type specification of the expressions right before the return-type of `FuntionLike`s is handled - I have not yet an idea where this place is though and whether thats a good idea :)

would be great to get a hint, where/how the actual param-out logic should be implemented.

### CODE REVIEW

File: tests/PHPStan/Analyser/data/param-out.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	$s .= "foo";
+}
+
+/**
+ * @template T of int
+ * @param-out T $s
+ */
+function genericFoo(mixed &$s): void
+{
+}
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/param-out.php]: played a bit with the psalm implementtion and found this interessting generic + param-out cases: https://psalm.dev/r/f30310f1f6

**Comment by ondrejmirtes** [comment]
In NodeScopeResolver there are places that are interested in `PassedByReference` - `@param-out` should be handled there.

Also - require your version of phpdoc-parser with the support added in composer.json, so that the pipeline can actually get green.

Also, you have to guard any call to new method on PhpDocNode with `method_exists` so that Rector can do its downgrading job. See: https://github.com/phpstan/phpstan-src/pull/1799#issuecomment-1269682926

**Comment by staabm** [comment]
thanks for the tips.

I tried testing the PR with the following code
```php
<?php

use function PHPStan\Testing\assertType;

class X {
	/**
	 * @param-out string $s
	 */
	function addFoo(?string &$s): void
	{
		if ($s === null) {
			$s = "hello";
		}
		$s .= "foo";
	}
}

function foo1(?string $s) {
	assertType('string|null', $s);
	$x = new X();
	$x->addFoo($s);
	assertType('string', $s);
}
```

but I currently struggle with php-doc resolving. 2 problems arise:
- ~~in `NodeScopeResolver->getParameterOutTypes(..)` the phpdoc `/** @param-out string $s */` gets always resolved to an empty-block from these lines~~ https://github.com/phpstan/phpstan-src/blob/49cd1319bf95d65a488fd51c543d93eeac049f8b/src/Type/FileTypeMapper.php#L111-L113
- ~~`FunctionReflection` does not yet contain a `getDocComment` method (like `MethodReflection` has)... it seems this is a missing part I need to implement with this PR?~~

**Comment by staabm** [comment]
I should write comments about my problems more often... sorry/thanks for rubberducking.. found the reason for doc-block resolving problem because I had mixed up caller and callee side of things - [48a2cf8](https://github.com/phpstan/phpstan-src/pull/1804/commits/48a2cf8ca5cd38e18c7c3b7b86679433d16925dd).

I think the PR works now as expected for basic method examples. open todo is for regular functions. I would implement a `getDocComment` method on `FunctionReflection` now, to get it working... if we agree on this

File: tests/PHPStan/Rules/PhpDoc/IncompatiblePhpDocTypeRuleTest.php [hunk 2]
```diff
@@ -233,4 +233,23 @@ public function testConditionalReturnType(): void
...
 
+	public function testParamOut(): void
+	{
+		$this->analyse([__DIR__ . '/data/param-out.php'], [
+			[
+				'PHPDoc tag @param-out references unknown parameter: $z',
+				23,
+			],
+			[
+				'PHPDoc tag @param-out for parameter $i with type string is incompatible with native type int.',
```
Comment (rvanvelzen) [inline tests/PHPStan/Rules/PhpDoc/IncompatiblePhpDocTypeRuleTest.php]: Why does the param-out type need to be compatible?

Comment (staabm) [reply tests/PHPStan/Rules/PhpDoc/IncompatiblePhpDocTypeRuleTest.php]: good catch. by-ref seems to be allowed to overwrite indepent from the parameters type
https://3v4l.org/W9C3X

fixed in the latest push

File: src/Analyser/NodeScopeResolver.php [hunk 3]
```diff
@@ -4098,6 +4108,53 @@ public function getPhpDocs(Scope $scope, Node\FunctionLike|Node\Stmt\Property $n
...
+		if ($resolvedPhpDoc === null) {
+			return [];
+		}
+
+		$phpDocParameterOutTypes = [];
+		foreach ($resolvedPhpDoc->getParamOutTags() as $paramName => $paramOutTag) {
+			$paramOutType = $paramOutTag->getType();
+			if ($scope->isInClass()) {
+				$paramOutType = $this->transformStaticType($scope->getClassReflection(), $paramOutType);
+			}
```
Comment (staabm) [inline src/Analyser/NodeScopeResolver.php]: I think this one here is not correct (copy paste leftover). need to find a proper testcase first..

File: src/Rules/PhpDoc/IncompatiblePhpDocTypeRule.php [hunk 4]
```diff
@@ -63,78 +65,107 @@ public function processNode(Node $node, Scope $scope): array
...
 		$errors = [];
 
-		foreach ($resolvedPhpDoc->getParamTags() as $parameterName => $phpDocParamTag) {
-			$phpDocParamType = $phpDocParamTag->getType();
-			if (!isset($nativeParameterTypes[$parameterName])) {
-				$errors[] = RuleErrorBuilder::message(sprintf(
-					'PHPDoc tag @param references unknown parameter: $%s',
-					$parameterName,
-				))->identifier('phpDoc.unknownParameter')->metadata(['parameterName' => $parameterName])->build();
+		foreach ([$resolvedPhpDoc->getParamTags(), $resolvedPhpDoc->getParamOutTags()] as $parameters) {
```
Comment (staabm) [inline src/Rules/PhpDoc/IncompatiblePhpDocTypeRule.php:72]: the more I work on the PR, the more I get the feeling that the param-put phpdoc validation should be put into a separate rule.. turns out most of the checks here are not necessary,  because we don't need to validate subtype/supertype correctness ..?

**Comment by staabm** [comment]
~~for some reason the PR as is failling after rebase, because I am running into~~ -> fixed by re-adding a recently added `use`-statement which got lost after manual merge conflict resolving

https://github.com/phpstan/phpstan-src/blob/f097ca90572d73f88ddbed6eb296eebfc5e72e7f/src/Analyser/NodeScopeResolver.php#L531-L533

**Comment by ondrejmirtes** [comment]
You updated a lot irrelevant deps. You need to run just `composer update phpstan/phpdoc-parser` instead.

**Comment by ondrejmirtes** [comment]
And now you get relevant build errors finally 😊

**Comment by staabm** [comment]
I had a look into the psalm test-suite for some more advanced use-cases... I will push them as a followup after we got the basic support landed

File: src/Analyser/MutatingScope.php [hunk 5]
```diff
@@ -2638,6 +2638,7 @@ public function enterFunction(
 		?bool $isPure = null,
 		bool $acceptsNamedArguments = true,
 		?Assertions $asserts = null,
+		?string $phpDocComment = null,
```
Comment (ondrejmirtes) [inline src/Analyser/MutatingScope.php:2647]: 1) All other parameters are specific about what they do here, I'd expect your added parameter to be specific about `@param-out`. Handling over the whole `$phpDocComment` smells like bad code.
2) In any case, you need to do the same for enterClassMethod.

File: src/Reflection/FunctionReflection.php [hunk 6]
```diff
@@ -34,4 +34,6 @@ public function isBuiltin(): bool;
 
 	public function getAsserts(): Assertions;
 
+	public function getDocComment(): ?string;
```
Comment (ondrejmirtes) [inline src/Reflection/FunctionReflection.php]: No, add a specific method for `@param-out`.

File: src/Analyser/NodeScopeResolver.php [hunk 7]
```diff
@@ -4177,6 +4190,49 @@ public function getPhpDocs(Scope $scope, Node\FunctionLike|Node\Stmt\Property $n
...
+		} else {
+			$file = $reflection->getFileName();
+		}
+
+		$resolvedPhpDoc = null;
+		$functionName = $reflection->getName();
+		$docComment = $reflection->getDocComment();
+
+		if ($docComment !== null) {
+			$resolvedPhpDoc = $this->fileTypeMapper->getResolvedPhpDoc(
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php]: `FunctionReflection|ExtendedMethodReflection` should already have a method for `@param-out` tags. But to be honest, it should probably be an information for a specific parameter on `ParameterReflectionWithPhpDocs` to be really clean. Something like `public function getOutType(): ?Type`.

Comment (ondrejmirtes) [reply src/Analyser/NodeScopeResolver.php]: You shouldn't definitely parse PHPDoc here again. That's a job for PhpClassReflectionExtension, `BetterReflectionProvider::getCustomFunction()` and `NativeFunctionReflectionProvider`.

File: src/Reflection/FunctionReflection.php (hunk 6 above)
Comment (ondrejmirtes) [reply src/Reflection/FunctionReflection.php]: Probably on ParameterReflectionWithPhpDocs.

File: test.php [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+function addValue(&$value) : void {
+	$value = 5;
+}
+
+function foo7() {
+	$foo = [];
+
+	addValue($foo["a"]);
+
+	\PHPStan\Testing\assertType('int', $foo["a"]);
```
Comment (ondrejmirtes) [inline test.php]: This file shouldn't be commited.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Also please test a scenario where `@param-out` on a method is inherited from a parent:

1) When the parameter names match.
2) When the parameters are renamed (but it should still be applied based on parameter order).
3) When the `@param-out` is overriden for the same parameter in the child class.

**Comment by ondrejmirtes** [comment]
Two more ideas:

* Support for `@param-out` in stubs. Should be tested in a similar fashion to this: https://github.com/phpstan/phpstan-src/blob/1.9.x/tests/PHPStan/Analyser/ConditionalReturnTypeFromMethodStubTest.php
* Support for conditional tags in `@param-out`.

**Comment by ondrejmirtes** [comment]
The stubs support needs to be implemented and fixed in multiple places. Essentially look at all the usages of StubPhpDocProvider::findMethodPhpDoc() and findFunctionPhpDoc():

* Functions coming from PHP
* Userland functions
* Methods coming from PHP
* Userland methods

File: src/Reflection/FunctionReflection.php (hunk 6 above)
Comment (staabm) [reply src/Reflection/FunctionReflection.php]: we have `getDocComment` on `ClassMemberReflection` and therefore I figured it would be good to also have it on `FunctionReflection` (for symmetry). having `getDocComment` on all relevant `*Reflection` classes is helpful to [build custom annotations which extension packages ](https://github.com/staabm/phpstan-dba/blob/7ce54d8d8938b08085bb63e760deb93b97e42104/src/PhpDoc/PhpDocUtil.php#L14-L39) sometimes need.

would you agree adding `getDocComment` on `FunctionReflection` as a separate PR meant for use in extensions?

I agree so for the `param-out` case a dedicated method would the way to go.

Comment (ondrejmirtes) [reply src/Reflection/FunctionReflection.php]: Yes please, add it on FunctionReflection, and fix it in PhpFunctionFromParserNodeReflection + PhpMethodFromParserNodeReflection. In a separate PR.

File: src/Analyser/MutatingScope.php (hunk 5 above)
Comment (staabm) [reply src/Analyser/MutatingScope.php:2647]: ~~actually its a leftover and will be dropped~~

**Comment by ondrejmirtes** [comment]
BTW this should also be taken advantage of in internal PHP function stubs (preg_match maybe?), feel free to look at Psalm stubs for these 😊

**Comment by ondrejmirtes** [comment]
Also https://github.com/phpstan/phpstan-src/pull/1674

File: src/Analyser/NodeScopeResolver.php [hunk 9]
```diff
@@ -3295,7 +3310,11 @@ private function processArgs(
 				if ($assignByReference) {
 					$argValue = $arg->value;
 					if ($argValue instanceof Variable && is_string($argValue->name)) {
-						$scope = $scope->assignVariable($argValue->name, new MixedType(), new MixedType());
+						$byRefType = new MixedType();
+						if (isset($parameters[$i]) && isset($paramOutTypes[$parameters[$i]->getName()])) {
+							$byRefType = $paramOutTypes[$parameters[$i]->getName()];
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php]: I'd do it a few lines above in the if/elseif to account for this edge-case: https://3v4l.org/QURqf (and with a test of course)

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Analyser/NodeScopeResolver.php (hunk 9 above)
Comment (staabm) [reply src/Analyser/NodeScopeResolver.php]: hmm.. does this mean the corresponding phpdoc would look like
`@param-out int ...$var`?

if so, we need another phpstan-parser PR which adds variadic into

https://github.com/phpstan/phpdoc-parser/blob/cc943678f73685b788d1030f6c6a4fbaef90a551/src/Ast/PhpDoc/ParamOutTagValueNode.php#L23

like we have it in 

https://github.com/phpstan/phpdoc-parser/blob/cc943678f73685b788d1030f6c6a4fbaef90a551/src/Ast/PhpDoc/ParamTagValueNode.php#L29

right?

Comment (ondrejmirtes) [reply src/Analyser/NodeScopeResolver.php]: I don't think so, `@param-out int $var` is fine

Comment (ondrejmirtes) [reply src/Analyser/NodeScopeResolver.php]: The `...` information is already in the native signature and/or in `@param`.

Comment (staabm) [reply src/Analyser/NodeScopeResolver.php]: I see, thanks. added a test and the proposed fix.

File: src/Rules/FunctionDefinitionCheck.php [hunk 10]
```diff
@@ -334,6 +334,17 @@ private function checkParametersAcceptor(
 
 					return $traverse($type);
 				});
+
+				if ($parameter instanceof ParameterReflectionWithPhpDocs && $parameter->getOutType() !== null) {
+					TypeTraverser::map($parameter->getOutType(), static function (Type $type, callable $traverse) use (&$templateTypes): Type {
+						if ($type instanceof TemplateType) {
+							unset($templateTypes[$type->getName()]);
```
Comment (ondrejmirtes) [inline src/Rules/FunctionDefinitionCheck.php]: This change wasn't needed for `@phpstan-assert` and doesn't make sense for `@param-out` either.

With a signature like this:

```php
/**
 * @template S of self
 * @param-out S $s
 */
function genericSelf(mixed &$s): void
```

The generics here don't do anything because it's the same as just writing `@param-out mixed $s`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Reflection/Php/PhpClassReflectionExtension.php [hunk 11]
```diff
@@ -518,6 +518,8 @@ private function createMethod(
 				$phpDocParameterTypes = [];
 				$phpDocReturnType = null;
 				$stubPhpDocPair = null;
+				$phpDocParameterOutTypes = [];
+				$phpDocParameterOutTypes = [];
```
Comment (ondrejmirtes) [inline src/Reflection/Php/PhpClassReflectionExtension.php:522]: duplicate line

File: src/Reflection/Php/PhpClassReflectionExtension.php [hunk 12]
```diff
@@ -598,7 +611,7 @@ private function createMethod(
 						}
 					}
 				}
-				$variants[] = $this->createNativeMethodVariant($methodSignature, $stubPhpDocParameterTypes, $stubPhpDocParameterVariadicity, $stubPhpDocReturnType, $phpDocParameterTypes, $phpDocReturnType, $phpDocParameterNameMapping);
+				$variants[] = $this->createNativeMethodVariant($methodSignature, $stubPhpDocParameterTypes, $stubPhpDocParameterVariadicity, $stubPhpDocReturnType, $phpDocParameterTypes, $phpDocReturnType, $phpDocParameterNameMapping, $phpDocParameterOutTypes, $phpDocParameterOutTypes);
```
Comment (ondrejmirtes) [inline src/Reflection/Php/PhpClassReflectionExtension.php]: duplicate

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Rules/FunctionDefinitionCheck.php (hunk 10 above)
Comment (staabm) [reply src/Rules/FunctionDefinitionCheck.php]: not sure I get your point. we need to unset the `$templateTypes` here, as we otherwise get a error below

> Template type S of function ParamOutTemplate\genericSelf() is not referenced in a parameter.

Comment (ondrejmirtes) [reply src/Rules/FunctionDefinitionCheck.php]: Yes, which is expected, because the generics for this function are useless. S will always be self.

**Comment by staabm** [comment]
I think it mostly works for basic cases. major case not working atm is phpdocs which should get inherited.

**Comment by ondrejmirtes** [comment]
Please do this too: https://github.com/phpstan/phpstan-src/pull/1804#issuecomment-1279928326 This part can give us valuable feedback about this feature working :)

**Comment by staabm** [comment]
alright: I included all psalm stubs which contain a `param-out` and picked a few tests out of the psalm codebase for those.

now we need to work out if the expectations are right and the remaining bugs :)

**Comment by staabm** [comment]
some errors I see regarding functions happen, because phpstan decides to use a function-variant from e.g. php8-stubs or function-map - which don't contain information about `param-out`.

should/can `param-out` information be merged from phpdoc stubs into the variants of other signature sources?

**Comment by staabm** [comment]
after sleeping a night over it, I came to the conclusion that the problems we see stems from the fact that I have implemented `getOutType(): ?Type` on `ParameterReflectionWithPhpDocs` and therefore parameter-signatures taken from php8-stubs or the resources/functionMap do not deliver this information.

I see 2 possible solutions:
A) implement `getOutType(): ?Type` on `ParameterReflection` 
B) implement `getParameterOutTypes(): Type[]` on `FunctionReflection|ExtendedMethodReflection`


for both solutions I would drop `getOutType(): ?Type` from `ParameterReflectionWithPhpDocs`.. wdyt?

**Comment by ondrejmirtes** [comment]
1) We can't, it'd be a BC break
2) That's not as nice

NativeMethodReflection already has ParametersAcceptorWithPhpDocs[].

NativeFunctionReflection has only ParametersAcceptor, but that can/should be adjusted to ParametersAcceptorWithPhpDocs. Please do that.

**Comment by staabm** [comment]
ok - I think we arrived at a point, where all newly added param-out tests pass.

next step: work thru regressions

**Comment by staabm** [comment]
the remaining 5 errors look a bit suprising:

```
There were 5 failures:

1) PHPStan\Analyser\LegacyNodeScopeResolverTest::testAssignInIf with data set #11 (PHPStan\Analyser\MutatingScope Object (...), 'matches', PHPStan\TrinaryLogic Object (...), 'mixed')
Type of variable $matches does not match the expected one.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'mixed'
+'(0 is 256 ? array<array{string, int<-1, max>}> : (0 is 512 ? array<string|null> : (0 is 768 ? array<array{null, -1}|array{string, int<0, max>}> : array<string>)))'

C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\LegacyNodeScopeResolverTest.php:923
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\LegacyNodeScopeResolverTest.php:753

2) PHPStan\Analyser\LegacyNodeScopeResolverTest::testAssignInIf with data set #18 (PHPStan\Analyser\MutatingScope Object (...), 'matches2', PHPStan\TrinaryLogic Object (...), 'mixed')
Type of variable $matches2 does not match the expected one.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'mixed'
+'(0 is 256 ? array<array{string, int<-1, max>}> : (0 is 512 ? array<string|null> : (0 is 768 ? array<array{null, -1}|array{string, int<0, max>}> : array<string>)))'

C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\LegacyNodeScopeResolverTest.php:923
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\LegacyNodeScopeResolverTest.php:753

3) PHPStan\Analyser\LegacyNodeScopeResolverTest::testAssignInIf with data set #20 (PHPStan\Analyser\MutatingScope Object (...), 'matches3', PHPStan\TrinaryLogic Object (...), 'mixed')
Type of variable $matches3 does not match the expected one.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'mixed'
+'(0 is 256 ? array<array{string, int<-1, max>}> : (0 is 512 ? array<string|null> : (0 is 768 ? array<array{null, -1}|array{string, int<0, max>}> : array<string>)))'

C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\LegacyNodeScopeResolverTest.php:923
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\LegacyNodeScopeResolverTest.php:753

4) PHPStan\Analyser\LegacyNodeScopeResolverTest::testAssignInIf with data set #21 (PHPStan\Analyser\MutatingScope Object (...), 'matches4', PHPStan\TrinaryLogic Object (...), 'mixed')
Type of variable $matches4 does not match the expected one.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'mixed'
+'(0 is 256 ? array<array{string, int<-1, max>}> : (0 is 512 ? array<string|null> : (0 is 768 ? array<array{null, -1}|array{string, int<0, max>}> : array<string>)))'

C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\LegacyNodeScopeResolverTest.php:923
C:\dvl\Workspace\phpstan-src-staabm\tests\PHPStan\Analyser\LegacyNodeScopeResolverTest.php:753

5) PHPStan\Analyser\LegacyNodeScopeResolverTest::testAssignInIf with data set #30 (PHPStan\Analyser\MutatingScope Object (...), 'ternaryMatches', PHPStan\TrinaryLogic Object (...), 'mixed')
Type of variable $ternaryMatches does not match the expected one.
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'mixed'
+'(0 is 256 ? array<array{string, int<-1, max>}> : (0 is 512 ? array<string|null> : (0 is 768 ? array<array{null, -1}|array{string, int<0, max>}> : array<string>)))'
```

to me it looks more or less like an unresolved conditional return type. maybe thats kind of the way how the `LegacyNodeScopeResolverTest` works, and I just need to update the expected values?

**Comment by staabm** [comment]
> to me it looks more or less like an unresolved conditional return type. maybe thats kind of the way how the `LegacyNodeScopeResolverTest` works, and I just need to update the expected values?

verified this theses by putting the code in question into a regular param-out test.
so, I just adapt the test expectations with [503a1d2](https://github.com/phpstan/phpstan-src/pull/1804/commits/503a1d25a342a6e21a21e1d205ef124924316b0e)

File: stubs/core.stub [hunk 13]
```diff
@@ -67,3 +67,166 @@ function base64_encode(string $string) : string {}
...
+
+/**
+ * @template T
+ * @template TArray as array<T>
+ *
+ * @param TArray $array
+ * @param callable(T,T):int $callback
+ * @param-out (TArray is non-empty-array ? non-empty-list<T> : list<T>) $array
+ */
+function usort(array &$array, callable $callback): bool
```
Comment (ondrejmirtes) [inline stubs/core.stub]: There are already stubs for some of those functions, so you need to update an existing stub instead of adding a new one.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
I fixed the problem with LegacyNodeScopeResolverTest, you can rebase and adjust those asserts again. https://github.com/phpstan/phpstan-src/commit/a83fdeb3f2eca8f081ea2bdd449434ebf2a9586a

File: stubs/core.stub (hunk 13 above)
Comment (staabm) [reply stubs/core.stub]: done

**Comment by staabm** [comment]
> you can rebase and adjust those asserts again

thank you. done

File: src/Analyser/NodeScopeResolver.php [hunk 14]
```diff
@@ -4039,7 +4070,7 @@ private function processNodesForTraitUse($node, ClassReflection $traitReflection
 	}
 
 	/**
-	 * @return array{TemplateTypeMap, Type[], ?Type, ?Type, ?string, bool, bool, bool, bool|null, bool, bool, string|null, Assertions, ?Type}
+	 * @return array{TemplateTypeMap, Type[], ?Type, ?Type, ?string, bool, bool, bool, bool|null, bool, bool, string|null, Assertions, ?Type, Type[]}
```
Comment (ondrejmirtes) [inline src/Analyser/NodeScopeResolver.php]: The type here could be more precise - `array<string, Type>`

File: src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php [hunk 15]
```diff
@@ -137,17 +140,22 @@ public function findFunctionReflection(string $functionName): ?NativeFunctionRef
...
+						$phpDocType ?? new MixedType(),
+						$type,
 						$parameterSignature->passedByReference(),
 						$parameterSignature->isVariadic(),
 						$parameterSignature->getDefaultValue(),
+						$phpDoc !== null ? NativeFunctionReflectionProvider::getParamOutTypeFromPhpDoc($parameterSignature->getName(), $phpDoc) : null,
 					);
 				}, $functionSignature->getParameters()),
 				$functionSignature->isVariadic(),
 				TypehintHelper::decideType($functionSignature->getReturnType(), $phpDoc !== null ? $this->getReturnTypeFromPhpDoc($phpDoc) : null),
```
Comment (ondrejmirtes) [inline src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php]: You haven't changed it but looks like on line 156 we could replace `$this->getReturnTypeFromPhpDoc($phpDoc)` with `$phpDocReturnType`?

File: src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php [hunk 16]
```diff
@@ -181,4 +189,14 @@ private function getReturnTypeFromPhpDoc(ResolvedPhpDocBlock $phpDoc): ?Type
...
 	}
 
+	private static function getParamOutTypeFromPhpDoc(string $paramName, ResolvedPhpDocBlock $stubPhpDoc): ?Type
+	{
+		foreach ($stubPhpDoc->getParamOutTags() as $name => $paramOutTag) {
+			if ($paramName === $name) {
+				return $paramOutTag->getType();
+			}
+		}
+		return null;
```
Comment (ondrejmirtes) [inline src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php:201]: Replace foreach with checking `$stubPhpDoc->getParamOutTags()[$paramName]` by array_key_exists?

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Super super nice! :) Just a few minor nitpicks :)

**Comment by staabm** [comment]
fixed and rebased. thank you

File: src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php (hunk 15 above)
Comment (staabm) [reply src/Reflection/SignatureMap/NativeFunctionReflectionProvider.php]: good catch. this line simplified a lot.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by staabm** [comment]
btw: I opened a new issue on php-src, to add/discuss `@param-out` directly into php-src-stubs

https://github.com/php/php-src/issues/9897

**Comment by ondrejmirtes** [comment]
@staabm Could you describe and open another issue for flag constants added to the stubs too? Like all the valid constants for `$flags` on json_encode (https://www.php.net/manual/en/function.json-encode.php) could be there.

**Comment by staabm** [comment]
> open another issue for flag constants added to the stubs too?

here we go: https://github.com/php/php-src/issues/9900
